apollo-federation/src/sources/graphql/federated_query_graph/builder.rs

```diff
@@ -9,9 +32,183 @@ pub(crate) struct FederatedQueryGraphBuilder;
+                    .get_type(subgraph_field.type_name.clone())
+                    .and_then(OutputTypeDefinitionPosition::try_from)
+                    .and_then(|ty| process_type(ty, subgraph_schema, builder, existing_types))
+                    .and_then(|next_index| {
+                        builder
+                            .add_concrete_field_edge(
+                                index,
+                                next_index,
+                                subgraph_field.field_name.clone(),
+                                IndexSet::new(),
+                                ConcreteFieldEdge {
+                                    subgraph_field,
+                                    requires_condition: None,
+                                }
+                                .into(),
+                            )
+                            .map(drop)
+                    })
+            })?,
+        Some(Either::Right(abstract_ty)) => subgraph_schema
+            .possible_runtime_types(abstract_ty.into())?
+            .into_iter()
+            .try_for_each(|ty| {
+                process_type(ty.clone().into(), subgraph_schema, builder, existing_types).and_then(
+                    |next_index| {
+                        builder
+                            .add_type_condition_edge(
+                                index,
+                                next_index,
+                                TypeConditionEdge::new(ty.into()).into(),
+                            )
+                            .map(drop)
+                    },
+                )
+            })?,
+        None => {}
+    }
+    Ok(index)
+}
+
+fn is_field_ignored(
+    field_definition_position: FieldDefinitionPosition,
+    subgraph_schema: &ValidFederationSchema,
+) -> bool {
+    if field_definition_position.field_name().starts_with("__") {
+        return true;
+    }
+
+    if let FieldDefinitionPosition::Object(position) = &field_definition_position {
+        let query_def = ObjectTypeDefinitionPosition {
+            type_name: name!("Query"),
+        };
+        let entities = query_def.field(name!("_entities"));
+        let service = query_def.field(name!("_service"));
+        if [entities, service].contains(position) {
+            return true;
+        }
     }
+
+    subgraph_schema
+        .subgraph_metadata()
+        .unwrap()
+        .external_metadata()
+        .is_external(&field_definition_position)
+        .unwrap_or_default()
+}
+
+fn is_source_entering_type_ignored(
+    output_type_definition_position: &OutputTypeDefinitionPosition,
+    subgraph_schema: &ValidFederationSchema,
+) -> bool {
+    let OutputTypeDefinitionPosition::Object(ty) = output_type_definition_position else {
+        return false;
+    };
+    let Ok(mut fields) = ty.field_positions(subgraph_schema.schema()) else {
+        return false;
+    };
+    fields.all(|field| is_field_ignored(field.into(), subgraph_schema))
```

Comment:
Similarly to the above, you'll want to change this function to return `Result<bool, FederationError>`, and if you're using `process_results()`, this line changes to e.g.:
```rust
fields.map(|field| is_field_ignored(field.into(), subgraph_schema))
    .process_results(|mut iter| iter.all(identity))
```